ulib/experimental/FStar.OrdSet.fsti

```diff
@@ -145,59 +181,158 @@ val fold (#a:eqtype) (#acc:Type) (#f:cmp a) (g:acc -> a -> acc) (init:acc) (s:or
 val map (#a #b:eqtype) (#fa:cmp a) (#fb:cmp b) (g:a -> b) (sa:ordset a fa)
   : Pure (ordset b fb)
     (requires (forall x y. x `fa` y ==> g x `fb` g y))
-    (ensures (fun sb -> True))
+    (ensures (fun sb -> (size sb <= size sa) /\                   
+                     (let sa = as_list sa in
+                      let sb = as_list sb in
+                      Cons? sb ==> Cons? sa /\ Cons?.hd sb == g (Cons?.hd sa))))
 
 val lemma_strict_subset_size (#a:eqtype) (#f:cmp a) (s1:ordset a f) (s2:ordset a f)
-  :Lemma (requires (strict_subset s1 s2))
-         (ensures  (subset s1 s2 /\ size s1 < size s2))
-   [SMTPat (strict_subset s1 s2)]
+  : Lemma (requires (strict_subset s1 s2))
+          (ensures  (subset s1 s2 /\ size s1 < size s2))
+    [SMTPat (strict_subset s1 s2)]
 
 val lemma_minus_mem (#a:eqtype) (#f:cmp a) (s1:ordset a f) (s2:ordset a f) (x:a)
-  :Lemma (requires True) (ensures (mem x (minus s1 s2) = (mem x s1 && not (mem x s2))))
-   [SMTPat (mem x (minus s1 s2))]
+  : Lemma (requires True) (ensures (mem x (minus s1 s2) = (mem x s1 && not (mem x s2))))
+    [SMTPat (mem x (minus s1 s2))]
+
+val lemma_strict_subset_exists_diff (#a:eqtype) (#f:cmp a) (s1:ordset a f) (s2:ordset a f) 
+  : Lemma (requires subset s1 s2)
+          (ensures (strict_subset s1 s2) <==> (exists x. (mem x s2 /\ not (mem x s1)))) 
+
+type condition a = a -> bool
+
+let inv #a (c: condition a) : (z:condition a{forall x. c x = not (z x)}) = fun x -> not (c x)
+
+val count (#a:eqtype) (#f: cmp a) (s: ordset a f) (c: condition a) : nat
```

Comment:
One idea you may consider is writing a general filter function with the spec (roughly):

```
val filter #a #f s c : Pure (ordset a f) (requires True) (ensures fun r -> as_list r == List.Tot.filter c (as_list s))
```

Then the lemmas here may be provable without knowing the internal representation of the ordset.

In general, it would also be nice to follow the FStar.Seq model, there is FStar.Seq.Base.fsti and FStar.Seq.Base.fst, which contains the basic definitions of sequences. Then there is a FStar.Seq.Properties.fsti/fst, which contain additional properties about sequences, that are derivable with just the interface of FStar.Seq.Base.

But doing this in a separate iteration, before we move the module out of experimental, is totally fine.